Improve rho diagnostic

EZoni opened the pull request:
This PR fixes two issues mentioned in #1037, namely:
1. the charge deposition diagnostics contained `rho` from the previous time step;
2. `rho` was computed at every time step whenever requested in the diagnostics.

The last improvement mentioned in #1037, namely making the new functor compute per-species charge deposition, can be addressed in another PR.

The variable `WarpX::plot_rho` has been removed from the WarpX class. A new functor `RhoFunctor` is defined and implemented in **Source/Diagnostics/ComputeDiagFunctors/RhoFunctor.H/.cpp**, following the implementation of similar functors, such as, for example, `DivEFunctor`.

I had to reset the benchmarks for the following tests: Langmuir_multi, Langmuir_multi_nodal, Langmuir_multi_single_precision, and Python_LaserAccelerationMR. These tests use all the FDTD solver, hence the `rho` values in the respective benchmarks were computed from the second-to-last iteration, because of the time staggering issue that this PR is fixing.

This PR should allow #1048 to be unblocked and merged, as the possible performance regression introduced there is now avoided by not depositing `rho` at every time step when requested in the diagnostics.

dpgrote commented on Source/Diagnostics/ComputeDiagFunctors/RhoFunctor.cpp:
```diff
@@ -0,0 +1,60 @@
+    rho_ptr = std::unique_ptr<amrex::MultiFab>
+              (new amrex::MultiFab( ba, warpx.DistributionMap(m_lev), 2*warpx.n_rz_azimuthal_modes-1, ng ));
+    rho_ptr = mypc.GetChargeDensity( m_lev );
+    amrex::MultiFab& rho = *rho_ptr;
```
I'll jump in. All that is needed here (instead of lines 23 to 35) is
`std::unique_ptr<amrex::MultiFab> rho = mypc.GetChargeDensity( m_lev );`
Then in the code below, use pointer dereferencing, e.g. `rho->boxArray()` and `*rho`.

dpgrote commented on Source/Diagnostics/ComputeDiagFunctors/RhoFunctor.cpp:
```diff
@@ -0,0 +1,60 @@
+    // mf_src should not be used, let's make sure it is null
+    AMREX_ALWAYS_ASSERT(mf_src == nullptr);
+    // Write only in one output component
+    AMREX_ALWAYS_ASSERT(ncomp == 1);
```
This assert should not be done since it is valid to request writing out all modes. See my suggested changes in FullDiagnostics.cpp.

dpgrote commented on Source/Diagnostics/FullDiagnostics.cpp:
```diff
@@ -6,6 +6,7 @@
+#include "ComputeDiagFunctors/RhoFunctor.H"
 #include "FlushFormats/FlushFormat.H"
 #include "FlushFormats/FlushFormatPlotfile.H"
 #include "FlushFormats/FlushFormatCheckpoint.H"
```
Some code needs to be added to the routine FullDiagnostics::AddRZModesToDiags to add the writing of the rho components. Duplicate the code that is there for divE, but writing rho. Note that the check for rho_requested can be simplified to by
`bool rho_requested = WarpXUtilStr::is_in(m_varnames, "rho");`
instead of the explicit loop over m_varnames.

ax3l replied:
:+1: looks addressed below now.

dpgrote commented:
@EZoni Thanks for the PR! See my suggestions.

dpgrote commented on Source/Diagnostics/ComputeDiagFunctors/RhoFunctor.cpp:
```diff
@@ -0,0 +1,52 @@
+    // Guard cell is set to 1 for generality. However, for a cell-centered
+    // output Multifab the guard-cell data is not needed, especially considering
+    // the operations performed in CoarsenIO::Coarsen
+    constexpr int ng = 1;
```
ng is not used here so this can be deleted.

dpgrote commented on Source/Diagnostics/FullDiagnostics.cpp:
```diff
@@ -194,6 +201,13 @@ FullDiagnostics::AddRZModesToDiags (int lev)
+                              m_crse_ratio, false, ncomp_multimodefab);
+        icomp += 1;
+        AddRZModesToOutputNames(std::string("rho"), ncomp_multimodefab);
+    }
```
Thanks for adding this!

dpgrote commented on Source/Diagnostics/ComputeDiagFunctors/RhoFunctor.cpp:
```diff
@@ -0,0 +1,52 @@
+      m_convertRZmodes2cartesian(convertRZmodes2cartesian)
+{
+    // mf_src should not be used, let's make sure it is null
+    AMREX_ALWAYS_ASSERT(mf_src == nullptr);
```
You can remove the mf_src argument since it isn't used for anything. This will be cleaner.

dpgrote commented on Regression/Checksum/benchmarks_json/Langmuir_multi_psatd_single_precision.json:
```diff
@@ -9,7 +9,7 @@
   },
   "lev=0": {
-    "Bx": 16.283409055462556,
+    "Bx": 16.302038904962248,
```
I would be careful about this change and make sure you understand why this update is needed. If this is due to a change in a different PR, then the change should have been part of that PR.

dpgrote commented:
@EZoni Thanks for the fixes. I have some more comments.

EZoni commented:
@dpgrote Thank you for all the suggestions. I think I implemented all of them now. I also reverted the Bx benchmark for the single precision PSATD test to its old value. I do not understand why this particular checksum test fails at all, given that I changed only the rho diagnostic, but I will try to look into this.

EZoni commented:
Despite reverting the Bx benchmark value, the checksum test did not fail, contrary to my expectations. So everything seems OK about that now. I see the test fail locally on my machine, which is why I reset that benchmark in the first place, but it could be something wrong on my side.

dpgrote commented:
@EZoni Is this ready for merge or do you want to wait for a review from someone else?

EZoni commented:
@dpgrote Thank you very much for your review!

Before merging, I would like to know @RevathiJambunathan's opinion about the question I asked in the PR description: do we have to be careful about back-transformed diagnostics and treat them as a special case, as suggested in @RemiLehe's [comment](https://github.com/ECP-WarpX/WarpX/pull/1048#issuecomment-633808537) to #1048?

I am not familiar with back-transformed diagnostics, so I don't know if this PR is introducing a bug somewhere where back-transformed diagnostics are used. The related tests pass, but they do not perform a checksum test, as far as I understand.

ax3l commented on Source/Diagnostics/ComputeDiagFunctors/RhoFunctor.H:
```diff
@@ -0,0 +1,47 @@
+     * \param[in] lev level of MultiFab
+     * \param[in] crse_ratio coarsening ratio for interpolation of field values
+     *                       from simulation MultiFabs to the output MultiFab mf_dst
+     * \param[in] convertRZmodes2cartesian
```
this params lacks a description :)

MaxThevenet commented:
Thanks for this PR!!
Indeed, we probably need a specific handling of back-transformed diagnostics as they also use rho, see `WarpX::GetCellCenteredData`, which contains
```c++
        const std::unique_ptr<MultiFab>& charge_density = mypc->GetChargeDensity(lev);
        AverageAndPackScalarField( *cc[lev], *charge_density, dmap[lev], dcomp, ng );
```
A few options:
1. Keep the old behavior of `plot_rho` just for BTD
2. remove the capability to plot rho in the BTD, at least temporarily.

I would be in favor of 1., because even if @RevathiJambunathan is reformatting BTD, the final implementation would have to do sth like plot_rho.
I will put the WIP tag back in the title of this PR.

RevathiJambunathan commented:
I agree. For now, we can have the old functionality just for BTD. I have an open PR and another follow-up with new BTD reformatting. I will include this functionality from RhoFunctor.cpp, in the new BTD.

EZoni commented:
@MaxThevenet @RevathiJambunathan 

Thank you for your comments and suggestions. I restored the old behavior of `plot_rho` for back-transformed diagnostics. In particular, the relevant changes are:
- Source/Diagnostics/Diagnostics.cpp: `warpx.setplot_rho(true)` (previously removed) is called again, but only when back-transformed diagnostics are used:
```cpp
// If user requests rho with back-transformed diagnostics, we set plot_rho=true
// and compute rho at each iteration
if (WarpXUtilStr::is_in(m_varnames, "rho") && WarpX::do_back_transformed_diagnostics) {
     warpx.setplot_rho(true);
}
```
- Source/Diagnostics/FullDiagnostics.cpp: the old functor `CellCenterFunctor` (previously removed) is called again, but only when back-transformed diagnostics are used:
```cpp
else if ( m_varnames[comp] == "rho" ){
    if ( WarpX::do_back_transformed_diagnostics ) {
#ifdef WARPX_USE_PSATD
        // rho_new is stored in component 1 of rho_fp when using PSATD
        amrex::MultiFab* rho_new = new amrex::MultiFab(*warpx.get_pointer_rho_fp(lev), amrex::make_alias, 1, 1);
        m_all_field_functors[lev][comp] = std::make_unique<CellCenterFunctor>(rho_new, lev, m_crse_ratio);
#else
        m_all_field_functors[lev][comp] = std::make_unique<CellCenterFunctor>(warpx.get_pointer_rho_fp(lev), lev, m_crse_ratio);
#endif
    }
    else {
        m_all_field_functors[lev][comp] = std::make_unique<RhoFunctor>(lev, m_crse_ratio);
    }
}
```
I thought that restoring the old functor in this case is necessary in order to avoid depositing the charge twice at every iteration. The new functor `RhoFunctor` is used in any other case, when back-transformed diagnostics are not used.
- Source/WarpX.cpp: the `MultiFab`s `rho_fp` and `rho_cp` are now allocated when both `plot_rho` and `do_back_transformed_diagnostics` are true. 

I will re-request your review, so that you can crosscheck this fix and provide feedback. Thank you.

RevathiJambunathan commented:
Thank you for this PR and for the changes @EZoni ! Looks good to me!

ax3l commented on Docs/source/visualization/advanced.rst:
```diff
@@ -17,7 +17,7 @@ Dump additional data
 
 In order to dump additional data in WarpX (mostly for debugging purpose), run the simulation
-with parameters
+with parameters (`warpx.plot_rho` only when using back-transformed diagnostics)
```
Markdown does verbatim with one backtick, rst takes two :)
```suggestion
with parameters (``warpx.plot_rho`` only when using back-transformed diagnostics)
```
